internal/querycoord/replica.go

```diff
@@ -112,3 +122,102 @@ func (rep *ReplicaInfos) Remove(replicaIds ...UniqueID) {
 		}
 	}
 }
+
+// ApplyBalancePlan applies balancePlan to replica nodes.
+func (rep *ReplicaInfos) ApplyBalancePlan(p *balancePlan, kv kv.MetaKv) error {
+	rep.globalGuard.Lock()
+	defer rep.globalGuard.Unlock()
+
+	var sourceReplica, targetReplica *milvuspb.ReplicaInfo
+
+	// check source and target replica ids are valid
+	if p.sourceReplica != invalidReplicaID {
+		replica, ok := rep.replicas[p.sourceReplica]
+		if !ok {
+			return errors.New("replica not found")
+		}
+		sourceReplica = proto.Clone(replica).(*milvuspb.ReplicaInfo)
+	}
+
+	if p.targetReplica != invalidReplicaID {
+		replica, ok := rep.replicas[p.targetReplica]
+		if !ok {
+			return errors.New("replica not found")
+		}
+		targetReplica = proto.Clone(replica).(*milvuspb.ReplicaInfo)
+	}
+
+	var replicasChanged []*milvuspb.ReplicaInfo
+
+	// generate ReplicaInfo to save to MetaKv
+	if sourceReplica != nil {
+		// remove node from replica node list
+		removeNodeFromReplica(sourceReplica, p.nodeID)
+		replicasChanged = append(replicasChanged, sourceReplica)
+	}
+	if targetReplica != nil {
+		// add node to replica
+		targetReplica.NodeIds = append(targetReplica.NodeIds, p.nodeID)
+		replicasChanged = append(replicasChanged, targetReplica)
+	}
+
+	// save to etcd first
+	if len(replicasChanged) > 0 {
+		data := make(map[string]string)
+
+		for _, info := range replicasChanged {
+			infoBytes, err := proto.Marshal(info)
+			if err != nil {
+				return err
+			}
+
+			key := fmt.Sprintf("%s/%d", ReplicaMetaPrefix, info.ReplicaID)
```

Comment:
Should Meta provide a method to save replica? Or just let insert() method do this

Reply:
This save operation may contain multiple replica info entries, temporary use dedicated code. We could refactor that if this situation changes.